In [3]:
from pathlib import Path
from IPython.display import display

ATTACK_DIR = Path("../exports/attack2-kube-apiserver")
ATTACK_SUFFIX = "attack2"

%run -i compare_logs.py

print("=== Resume ===")
display(summary)

print("\n=== Events per second (stats) ===")
display(eps_stats)

print("\n=== Resource x Verb ===")
display(resource_x_verb)

for field, df in results.items():
    print(f"\n=== {field} ===")
    display(df)

        métrique  baseline  attack  variation
0   total_events      2089    4784       2695
1  patch_on_pods         0       0          0
=== Resume ===


,métrique,baseline,attack,variation
0,total_events,2089,4784,2695
1,patch_on_pods,0,0,0



=== Events per second (stats) ===


,métrique,baseline,attack,variation
0,max_events_par_sec,25.0,64.00,39.00
1,moyenne_events_par_sec,7.6,29.53,21.93



=== Resource x Verb ===


,resource,verb,count_baseline,count_attack,variation
0,leases,update,710,1418,708
1,nodes,get,87,348,261
2,endpointslices,get,58,120,62
3,endpoints,get,58,120,62
4,clusterinformations,get,60,120,60
...,...,...,...,...,...
62,clusterinformations,watch,3,0,-3
63,podtemplates,watch,3,0,-3
64,poddisruptionbudgets,watch,3,0,-3
65,serviceaccounts,watch,6,0,-6



=== verb ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,get,1063,2308,212.6,461.6,1245,249.0
1,update,710,1418,142.0,283.6,708,141.6
2,watch,309,690,61.8,138.0,381,76.2
3,patch,3,12,0.6,2.4,9,1.8
4,list,4,8,0.8,1.6,4,0.8



=== objectRef_resource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,leases,713,1430,142.6,286.0,717,143.4
1,nodes,108,414,21.6,82.8,306,61.2
2,endpointslices,70,144,14.0,28.8,74,14.8
3,endpoints,64,126,12.8,25.2,62,12.4
4,clusterinformations,63,120,12.6,24.0,57,11.4
5,services,28,62,5.6,12.4,34,6.8
6,configmaps,21,54,4.2,10.8,33,6.6
7,csidrivers,9,30,1.8,6.0,21,4.2
8,pods,12,30,2.4,6.0,18,3.6
9,runtimeclasses,9,24,1.8,4.8,15,3.0



=== objectRef_namespace ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,kube-system,643,1284,128.6,256.8,641,128.2
1,default,116,240,23.2,48.0,124,24.8
2,kube-node-lease,88,176,17.6,35.2,88,17.6
3,attacker-ns,0,6,0.0,1.2,6,1.2
4,safe-ns,0,6,0.0,1.2,6,1.2



=== objectRef_subresource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,status,3,12,0.6,2.4,9,1.8



=== responseStatus_code ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,200,990,2212,198.0,442.4,1222,244.4



=== authorization_decision ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,allow,990,2212,198.0,442.4,1222,244.4



=== authorization_reason ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,"RBAC: allowed by ClusterRoleBinding ""system:pu...",330,660,66.0,132.0,330,66.0
1,"RBAC: allowed by ClusterRoleBinding ""system:ku...",236,476,47.2,95.2,240,48.0
2,"RBAC: allowed by ClusterRoleBinding ""system:ku...",161,340,32.2,68.0,179,35.8
3,"RBAC: allowed by ClusterRoleBinding ""system:di...",65,118,13.0,23.6,53,10.6
4,"RBAC: allowed by ClusterRoleBinding ""calico-ku...",30,68,6.0,13.6,38,7.6
5,"RBAC: allowed by ClusterRoleBinding ""calico-no...",14,32,2.8,6.4,18,3.6
6,"RBAC: allowed by ClusterRoleBinding ""kubeadm:n...",14,24,2.8,4.8,10,2.0
7,"RBAC: allowed by ClusterRoleBinding ""system:pu...",5,22,1.0,4.4,17,3.4
8,"RBAC: allowed by ClusterRoleBinding ""system:co...",6,12,1.2,2.4,6,1.2
9,"RBAC: allowed by ClusterRoleBinding ""system:vo...",5,4,1.0,0.8,-1,-0.2



=== sourceIPs ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,10.10.10.10,1572,3238,314.4,647.6,1666,333.2
1,::1,210,420,42.0,84.0,210,42.0
2,10.10.10.11,83,276,16.6,55.2,193,38.6
3,192.168.62.165,0,252,0.0,50.4,252,50.4
4,10.10.10.12,89,232,17.8,46.4,143,28.6
5,192.168.62.166,0,18,0.0,3.6,18,3.6
6,10.0.2.15,6,0,1.2,0.0,-6,-1.2
7,192.168.62.160,9,0,1.8,0.0,-9,-1.8
8,192.168.62.161,120,0,24.0,0.0,-120,-24.0



=== user_groups ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:authenticated,1429,3116,285.8,623.2,1687,337.4
1,system:unauthenticated,660,1320,132.0,264.0,660,132.0
2,system:nodes,217,656,43.4,131.2,439,87.8
3,system:serviceaccounts,251,514,50.2,102.8,263,52.6
4,system:serviceaccounts:kube-system,251,514,50.2,102.8,263,52.6
5,system:masters,210,420,42.0,84.0,210,42.0



=== user_username ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:anonymous,660,1320,132.0,264.0,660,132.0
1,system:kube-controller-manager,428,862,85.6,172.4,434,86.8
2,system:kube-scheduler,323,664,64.6,132.8,341,68.2
3,system:apiserver,210,420,42.0,84.0,210,42.0
4,system:serviceaccount:kube-system:calico-kube-...,120,252,24.0,50.4,132,26.4
5,system:node:k8s-w1,71,222,14.2,44.4,151,30.2
6,system:node:k8s-cp,72,220,14.4,44.0,148,29.6
7,system:node:k8s-w2,74,214,14.8,42.8,140,28.0
8,system:serviceaccount:kube-system:generic-garb...,40,80,8.0,16.0,40,8.0
9,system:serviceaccount:kube-system:resourcequot...,40,80,8.0,16.0,40,8.0



=== userAgent ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,kube-probe/1.30,660,1320,132.0,264.0,660,132.0
1,kubelet/v1.30.14 (linux/amd64) kubernetes/9e18483,217,656,43.4,131.2,439,87.8
2,kube-controller-manager/v1.30.14 (linux/amd64)...,296,592,59.2,118.4,296,59.2
3,kube-scheduler/v1.30.14 (linux/amd64) kubernet...,296,592,59.2,118.4,296,59.2
4,kube-apiserver/v1.30.14 (linux/amd64) kubernet...,210,420,42.0,84.0,210,42.0
5,kube-controller-manager/v1.30.14 (linux/amd64)...,96,198,19.2,39.6,102,20.4
6,Go-http-client/2.0,72,162,14.4,32.4,90,18.0
7,kube-controllers/v0.0.0 (linux/amd64) kubernet...,60,132,12.0,26.4,72,14.4
8,kube-controller-manager/v1.30.14 (linux/amd64)...,40,80,8.0,16.0,40,8.0
9,kube-controller-manager/v1.30.14 (linux/amd64)...,40,80,8.0,16.0,40,8.0


## Results summary

| Metric | Baseline | Attack | Variation | Factor |
|---|---|---|---|---|
| Total audit events | 2089 | 4784 | +2695 | x2.3 |
| Events/sec (average) | 7.6 | 29.53 | +21.93 | x3.9 |
| Events/sec (peak) | 25 | 64 | +39 | x2.6 |
| `authorization.decision: allow` | 990 | 2212 | +1222 | x2.2 |
| `patch` on `pods` | 0 | 0 | 0 | - |

| Verb | Baseline | Attack | Factor |
|---|---|---|---|
| get | 1063 | 2308 | x2.2 |
| update | 710 | 1418 | x2.0 |
| watch | 309 | 690 | x2.2 |
| patch | 3 | 12 | x4.0 |

| Namespace | Baseline | Attack |
|---|---|---|
| kube-system | 643 | 1284 |
| default | 116 | 240 |
| kube-node-lease | 88 | 176 |
| attacker-ns | 0 | 6 |

---

## Analysis of results

The dominant pattern in this capture is a broad, roughly uniform amplification of existing baseline activity rather than the appearance of a new, concentrated signal. Every verb grows by close to the same factor (get x2.2, update x2.0, watch x2.2), and the same holds for `userAgent`: `kube-probe` (660 to 1320), `kubelet` (217 to 656), `kube-controller-manager` leader-election (296 to 592), `kube-scheduler` leader-election (296 to 592), and `kube-apiserver` itself (210 to 420) all scale up by roughly the same x2 factor, with no new account appearing that was absent in the baseline. The `authorization.reason` breakdown confirms this: every RBAC grant already present in the baseline (`system:public-info-viewer`, `system:kube-controller-manager`, `system:kube-scheduler`, `calico-kube-controllers`, and so on) simply increases in volume, without any new ClusterRoleBinding being exercised.

This uniform x2 inflation is consistent with the nf_conntrack saturation mechanism driving this attack: once the connection tracking table fills, packets are dropped randomly, which breaks existing `watch` connections held by these same clients and forces them to reconnect and resynchronize, typically through a full `list` call rather than resuming an efficient `watch`. This produces more API calls from each client, not fewer, even though the clients themselves are not the ones under attack, which explains why the increase shows up broadly across nearly every existing account rather than as a spike concentrated on one of them.

Within this uniform growth, `update` has the smallest factor of the four verbs (x2.0, against x2.2 for `get` and `watch`), while most of the reconnection/resync traffic described above falls under `get`/`list`/`watch`. These read operations are typically served from the kube-apiserver's in-memory watch cache rather than by querying etcd directly, so this traffic increase does not necessarily translate into a proportional increase in etcd load. This is consistent with a separate measurement from the same attack window, where etcd CPU usage falls rather than rises during the attack, even as kube-apiserver CPU and overall client-facing traffic both increase.

`objectRef.namespace` shows the same pattern: the growth is concentrated in `kube-system` (643 to 1284), `default` (116 to 240), and `kube-node-lease` (88 to 176), the namespaces that host normal control plane chatter, while `attacker-ns` stays almost untouched (0 to 6). This is consistent with the attack targeting the `/healthz` endpoint of the `kube-apiserver` directly: `/healthz` is a non-resource URL with no associated `objectRef`, so the attacker's own requests do not themselves populate the `objectRef.resource`/`objectRef.namespace` breakdown. What we observe instead is the knock-on effect described above: existing clients reconnecting and resyncing more often, which inflates their own normal traffic rather than introducing a distinct new signal.

`patch_on_pods` stays at 0 in both scenarios, confirming that this attack does not touch pod status at all, unlike mechanisms built on the readiness probe. `objectRef.subresource: status` barely moves (3 to 12, a negligible absolute count), reinforcing that the attack's footprint is not expressed through subresource state churn.

`sourceIPs` shows one notable new entry, `192.168.62.165` (0 to 252), alongside the growth of existing source IPs (`10.10.10.10`: 1572 to 3238, `::1`: 210 to 420, `10.10.10.11`: 83 to 276). The loopback address `::1` doubling is itself a side effect worth noting: it reflects the `kube-apiserver`'s own internal health-check traffic increasing under load, not external attacker activity.

---

## Detectability analysis

**1. No single field stands out sharply.** Unlike a pattern driven by a specific resource or a new controller account, this attack's signature is diffuse: almost every verb, namespace, and userAgent scales by roughly the same factor. A detection rule keyed on any single field (a specific resource, a specific account) would likely miss this attack, since nothing in the baseline composition qualitatively changes, only its volume.

**2. Aggregate rate is the most reliable indicator available.** `events/sec` average grows x3.9 and peak x2.6. A threshold on overall event rate, independent of which field is driving it, is the most direct way to catch this kind of attack in these logs.

**3. A new source IP appearing at volume.** `192.168.62.165` goes from 0 to 252 occurrences and was entirely absent from the baseline. Tracking new source IPs crossing a volume threshold is a usable secondary indicator, though weaker than the account-based indicators available for other attack types, since it depends on IP visibility and stability.

### Detectability limitations to note

- **The attack's own requests are largely invisible in `objectRef` fields.** Because `/healthz` is a non-resource endpoint, the bulk of the malicious traffic does not appear in `objectRef.resource` or `objectRef.namespace` at all. Detection based on these fields would only catch the secondary effect (other clients retrying more), not the attack itself.
- **No new RBAC grant or controller account appears.** Every `authorization.reason` entry present in the attack was already present in the baseline, just at higher volume. A detector relying on new/unseen accounts or roles, effective for other attack types, would not trigger here.
- **The signal is proportionally weaker than the raw event count suggests.** A x2.3 increase in `total_events` is modest compared to the volume spikes seen in attacks that create new objects or drive status churn, which means a threshold tuned to catch those attacks might be set too high to catch this one reliably.